src/libraries/System.Private.CoreLib/src/System/SpanHelpers.Char.cs

```diff
@@ -71,63 +71,200 @@ public static unsafe int SequenceCompareTo(ref char first, int firstLength, ref
+                        offset += LocateFirstFoundChar(matches);
+                        goto OffsetDifference;
+
+                    } while (lengthToExamine > offset);
                 }
             }
 
 #if BIT64
-            if ((byte*)minLength >= (byte*)(i + sizeof(int) / sizeof(char)))
+            ulong difference;
+            while (minLength >= (offset + sizeof(ulong) / sizeof(char)))
             {
-                if (Unsafe.ReadUnaligned<int>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref first, i))) ==
-                    Unsafe.ReadUnaligned<int>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref second, i))))
+                ulong firstLong = Unsafe.ReadUnaligned<ulong>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref first, (IntPtr)offset)));
+                ulong secondLong = Unsafe.ReadUnaligned<ulong>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref second, (IntPtr)offset)));
+
+                difference = firstLong ^ secondLong;
+                if (difference != 0)
                 {
-                    i += sizeof(int) / sizeof(char);
+                    goto LongDifference;
                 }
+
+                offset += sizeof(ulong) / sizeof(char);
             }
 #endif
+            if (minLength >= (offset + sizeof(int) / sizeof(char)))
+            {
+                if (Unsafe.ReadUnaligned<int>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref first, (IntPtr)offset))) ==
+                    Unsafe.ReadUnaligned<int>(ref Unsafe.As<char, byte>(ref Unsafe.Add(ref second, (IntPtr)offset))))
+                {
+                    offset += sizeof(int) / sizeof(char);
+                }
+            }
 
-            while ((byte*)i < (byte*)minLength)
+            int result = 0;
+            while (offset < minLength)
             {
-                int result = Unsafe.Add(ref first, i).CompareTo(Unsafe.Add(ref second, i));
+                result = Unsafe.Add(ref first, (IntPtr)offset).CompareTo(Unsafe.Add(ref second, (IntPtr)offset));
                 if (result != 0)
-                    return result;
-                i += 1;
+                    goto ResultDifference;
+                offset++;
             }
 
         Equal:
-            return lengthDelta;
+            // At this point, we have compared all the characters in at least one string.
+            // The longer string will be larger; if they are the same length this will return 0.
+            return firstLength - secondLength;
+
+#if BIT64
+        LongDifference:
+            // Find bit offset of first difference and add to current offset,
+            // convert the bit position to char position by dividing by (2 * 8) for chars (shift right by 4)
+            offset += BitOperations.TrailingZeroCount(difference) >> 4;
+#endif
+        OffsetDifference:
+            result = Unsafe.Add(ref first, (IntPtr)offset).CompareTo(Unsafe.Add(ref second, (IntPtr)offset));
+            Debug.Assert(result != 0);
```

Comment:
This is an invalid `Debug.Assert`. The span contents could have changed while the comparison was taking place, so it's possible (though unlikely) that we'll get a 0 here.